src/agent.rs

```diff
@@ -40,34 +41,55 @@ pub struct Agent {
     /// Copied into each request of this agent.
     pub(crate) headers: Vec<Header>,
     /// Reused agent state for repeated requests from this agent.
-    pub(crate) state: Arc<Mutex<Option<AgentState>>>,
+    pub(crate) state: Arc<Mutex<AgentState>>,
```

Comment:
I think the resolver might better sit on `Agent` than `AgentState`. That would avoid rewriting this optional.

That would mean it needs to be also `Send + Sync`, we could also require it to extend `Debug`, to continue allow `#[derive(Debug)]`

Since we're not going to deal with `Resolver` that don't have these bounds, I think for simplicity we could just jam all the requirement into the trait itself.

```rust
pub trait Resolver: Debug + Send + Sync + 'static {
  ...
}

pub struct Agent {
   …
  pub(crate) resolver: Arc<dyn Resolver>,
}
```

Reply:
Hmm, so having the resolver on the Agent, would that imply that the resolver will be copied to the request when making one?

The reason I chose not to, was to avoid extra Arc::cloneing, but perhaps that's a low price to pay.

Reply:
I see what you mean. `Unit` has `AgentState` already 🤔